File: rd-net/RdFramework/Impl/ProtocolContexts.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Diagnostics.CodeAnalysis;
using System.Threading;
using JetBrains.Collections;
using JetBrains.Collections.Viewable;
using JetBrains.Diagnostics;
using JetBrains.Lifetimes;
using JetBrains.Rd.Base;
using JetBrains.Serialization;
using JetBrains.Util;

namespace JetBrains.Rd.Impl
{
  /// <summary>
  /// This class handles RdContext on protocol level. It tracks existing contexts and allows access to their value sets (when present)
  /// </summary>
  public class ProtocolContexts : RdReactiveBase
  {
    private readonly CopyOnWriteList<ISingleContextHandler> myCounterpartHandlers = new();
    private readonly CopyOnWriteList<ISingleContextHandler> myHandlersToWrite = new();
    private readonly IViewableList<ISingleContextHandler> myHandlerOrder = new ViewableList<ISingleContextHandler>();
    private readonly ConcurrentDictionary<RdContextBase, ISingleContextHandler> myHandlersMap = new();
    private readonly object myOrderingLock = new();
    private readonly ThreadLocal<bool> mySendWithoutContexts = new(() => false);


    internal readonly struct SendWithoutContextsCookie : IDisposable
    {
      private readonly ProtocolContexts myContexts;
      private readonly bool myPrevValue;

      public SendWithoutContextsCookie(ProtocolContexts contexts)
      {
        myContexts = contexts;
        myPrevValue = contexts.mySendWithoutContexts.Value;
        contexts.mySendWithoutContexts.Value = true;
      }

      public void Dispose()
      {
        myContexts.mySendWithoutContexts.Value = myPrevValue;
      }
    }

    private readonly SerializationCtx mySerializationCtx;

    internal SendWithoutContextsCookie CreateSendWithoutContextsCookie() => new SendWithoutContextsCookie(this);
    public bool IsSendWithoutContexts => mySendWithoutContexts.Value;

    public ProtocolContexts(SerializationCtx serializationCtx)
    {
      Async = true;
      mySerializationCtx = serializationCtx;
    }

    public ICollection<RdContextBase> RegisteredContexts => myHandlersMap.Keys;

    internal ISingleContextHandler<T> GetHandlerForContext<T>(RdContext<T> context)
    {
      return (ISingleContextHandler<T>) myHandlersMap[context];
    }

    public override void OnWireReceived(IProtocol proto, SerializationCtx ctx, UnsafeReader reader, IRdWireableDispatchHelper dispatchHelper)
    {
      var contextBase = RdContextBase.Read(mySerializationCtx, reader);

      contextBase.RegisterOn(this);

      myCounterpartHandlers.Add(myHandlersMap[contextBase]);
    }

    private void DoAddHandler<T>(RdContext<T> context, ISingleContextHandler<T> handler)
    {
      if (myHandlersMap.TryAdd(context, handler))
      {
        context.RegisterOn(mySerializationCtx.Serializers);
        lock (myOrderingLock)
          myHandlerOrder.Add(handler);
      }
    }

    private void PreBindHandler(Lifetime lifetime, string key, ISingleContextHandler handler, IProtocol proto)
    {
      if (handler is RdBindableBase bindableHandler)
      {
        bindableHandler.RdId = proto.Identities.Mix(RdId, key);
        bindableHandler.PreBind(lifetime, this, key);
      }
    }

    private void BindHandler(ISingleContextHandler handler)
    {
      if (handler is RdBindableBase bindableHandler)
      {
        using (CreateSendWithoutContextsCookie())
          bindableHandler.Bind();
      }
    }


    private void SendContextToRemote(RdContextBase context)
    {
      var wire = TryGetProto()?.Wire;
      if (wire == null)
        return;

      using(CreateSendWithoutContextsCookie())
        wire.Send(RdId, writer =>
        {
          RdContextBase.Write(mySerializationCtx, writer, context);
        });
    }

    private void EnsureHeavyHandlerExists<T>(RdContext<T> context)
    {
      if (Mode.IsAssertion) Assertion.Assert(context.IsHeavy, "key.IsHeavy");
      if (!myHandlersMap.ContainsKey(context))
        DoAddHandler(context, new HeavySingleContextHandler<T>(context, this));
    }

    private void EnsureLightHandlerExists<T>(RdContext<T> context)
    {
      if (Mode.IsAssertion) Assertion.Assert(!context.IsHeavy, "!key.IsHeavy");
      if (!myHandlersMap.ContainsKey(context))
        DoAddHandler(context, new LightSingleContextHandler<T>(context));
    }

    /// <summary>
    /// Get a value set for a given key. The values are local relative to transform
    /// </summary>
    public IAppendOnlyViewableConcurrentSet<T> GetValueSet<T>(RdContext<T> context) where T : notnull
    {
      if (Mode.IsAssertion) Assertion.Assert(context.IsHeavy, "Only heavy keys have value sets, key {0} is light", context.Key);
      return ((HeavySingleContextHandler<T>) GetHandlerForContext(context)).LocalValueSet;
    }

    /// <summary>
    /// Registers a context to be used with this context handler. Must be invoked on protocol's scheduler
    /// </summary>
    public void RegisterContext<T>(RdContext<T> context)
    {
      if (myHandlersMap.ContainsKey(context)) return;
      if(context.IsHeavy)
        EnsureHeavyHandlerExists(context);
      else
        EnsureLightHandlerExists(context);
    }


    protected override void PreInit(Lifetime lifetime, IProtocol proto)
    {
      base.PreInit(lifetime, proto);

      lock (myOrderingLock)
      {
        myHandlerOrder.View(lifetime, (handlerLt, _, handler) =>
        {
          PreBindHandler(handlerLt, handler.ContextBase.Key, handler, proto);
        });
      }

      proto.Wire.Advise(lifetime, this);
    }

    protected override void Init(Lifetime lifetime, IProtocol proto, SerializationCtx ctx)
    {
      base.Init(lifetime, proto, ctx);
      lock (myOrderingLock)
      {
        myHandlerOrder.View(lifetime, (handlerLt, _, handler) =>
        {
          BindAndSendHandler(handler);
        });
      }

    }

    /// <summary>
    /// Reads context values from a message, sets current context to them, and returns a cookie to restore previous context
    /// </summary>
    internal MessageContext ReadContextsIntoCookie(UnsafeReader reader)
    {
      var numContextValues = reader.ReadShort();
      if (numContextValues == 0)
        return default;

      var handlers = myCounterpartHandlers;
      if (Mode.IsAssertion) Assertion.Assert(numContextValues <= handlers.Count, "We know of {0} other side keys, received {1} instead", handlers.Count, numContextValues);

      var values = new object[numContextValues];
      for (var i = 0; i < numContextValues; i++)
        values[i] = handlers[i].ReadValueBoxed(mySerializationCtx, reader);

      return new MessageContext(values, handlers.GetStorageUnsafe());
    }

    internal readonly ref struct MessageContextCookie
    {
      private readonly IDisposable[] myDisposables;

      public MessageContextCookie(IDisposable[] disposables)
      {
        myDisposables = disposables;
      }

      public void Dispose()
      {
        if (myDisposables is { } disposables)
        {
          foreach (var disposable in disposables)
            disposable.Dispose();
        }
      }
    }

    internal readonly struct MessageContext
    {
      private readonly object[] myValues;
      private readonly ISingleContextHandler[] myHandlers;

      public MessageContext(object[] values, ISingleContextHandler[] handlers)
      {
        myValues = values;
        myHandlers = handlers;
      }

      public MessageContextCookie UpdateCookie()
      {
        if (myHandlers == null)
          return default;

        var disposables = new IDisposable[myValues.Length];
        for (var i = 0; i < myValues.Length; i++)
          disposables[i] = myHandlers[i].ContextBase.UpdateValueBoxed(myValues[i]);

        return new MessageContextCookie(disposables);
      }
    }

    /// <summary>
    /// Writes the current context values
    /// </summary>
    [SuppressMessage("ReSharper", "InconsistentlySynchronizedField", Justification = "sync is for atomicity of write/send pairs, not access")]
    public void WriteContexts(UnsafeWriter writer)
    {
      if (IsSendWithoutContexts)
      {
        WriteEmptyContexts(writer);
        return;
      }
// all handlers in myHandlersToWrite have been sent to the remote side
      var count = myHandlersToWrite.Count;
      writer.WriteInt16((short) count);
      for (var i = 0; i < count; i++)
        myHandlersToWrite[i].WriteValue(mySerializationCtx, writer);
    }

    /// <summary>
    /// Adds current values of registered contexts to their respective value sets without writing them to the wire
    /// </summary>
    [SuppressMessage("ReSharper", "InconsistentlySynchronizedField", Justification = "sync is for atomicity of write/send pairs, not access")]
    public void RegisterCurrentValuesInValueSets()
    {
      var count = myHandlerOrder.Count;
      for (var i = 0; i < count; i++)
        myHandlerOrder[i].RegisterValueInValueSet();
    }

    /// <summary>
    /// Writes an empty context
    /// </summary>
    public static void WriteEmptyContexts(UnsafeWriter writer)
    {
      writer.WriteInt16((short) 0);
    }

    private void BindAndSendHandler(ISingleContextHandler handler)
    {
      SendContextToRemote(handler.ContextBase);
      BindHandler(handler);
      // add the handler to myHandlersToWrite only after sending the context to remote
      myHandlersToWrite.Add(handler);
    }
  }
}

File: rd-kt/rd-framework/src/main/kotlin/com/jetbrains/rd/framework/impl/ProtocolContexts.kt
package com.jetbrains.rd.framework.impl

import com.jetbrains.rd.framework.*
import com.jetbrains.rd.framework.base.*
import com.jetbrains.rd.framework.base.ISingleContextHandler
import com.jetbrains.rd.util.ConcurrentHashMap
import com.jetbrains.rd.util.CopyOnWriteArrayList
import com.jetbrains.rd.util.Sync
import com.jetbrains.rd.util.assert
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.IAppendOnlyViewableConcurrentSet
import com.jetbrains.rd.util.reactive.ViewableList
import com.jetbrains.rd.util.reflection.threadLocal
import com.jetbrains.rd.util.reflection.usingValue

/**
 * This class handles RdContext on protocol level. It tracks existing contexts and allows access to their value sets (when present)
 */
class ProtocolContexts(val serializationCtx: SerializationCtx) : RdReactiveBase() {
    private val counterpartHandlers = CopyOnWriteArrayList<ISingleContextHandler<*>>()
    private val handlersToWrite = CopyOnWriteArrayList<ISingleContextHandler<*>>()
    private val myHandlerOrder = ViewableList<ISingleContextHandler<*>>()
    private val handlersMap = ConcurrentHashMap<RdContext<*>, ISingleContextHandler<*>>()
    private val myOrderingsLock = Any() // used to protect writes to myKeyHandlersOrder

    internal var isSendWithoutContexts by threadLocal { false }
    internal fun <T> sendWithoutContexts(block: () -> T) = this::isSendWithoutContexts.usingValue(true, block)

    override fun deepClone(): IRdBindable {
        error("This may not be cloned")
    }

    val registeredContexts: Collection<RdContext<*>>
        get() = handlersMap.keys

    override fun preInit(lifetime: Lifetime, proto: IProtocol) {
        super.preInit(lifetime, proto)

        Sync.lock(myOrderingsLock) {
            myHandlerOrder.view(lifetime) { handlerLifetime, _, handler ->
                preBindHandler(handlerLifetime, handler, handler.context.key, proto)
            }
        }

        proto.wire.advise(lifetime, this)
    }

    override fun init(lifetime: Lifetime, proto: IProtocol, ctx: SerializationCtx) {
        super.init(lifetime, proto, ctx)
        Sync.lock(myOrderingsLock) {
            myHandlerOrder.view(lifetime) { handlerLifetime, _, handler ->
                bindHandler(handler)
                sendContextToRemote(handler.context)

                // add the handler to myHandlersToWrite only after sending the context to remote
                handlersToWrite.add(handler)
            }
        }
    }

    override fun onWireReceived(proto: IProtocol, buffer: AbstractBuffer, ctx: SerializationCtx, dispatchHelper: IRdWireableDispatchHelper) {
        val context = RdContext.read(serializationCtx, buffer)
        ensureContextHandlerExists(context)
        counterpartHandlers.add(handlersMap[context]!!)
    }

    private inline fun doAddHandler(context: RdContext<*>, factory: () -> ISingleContextHandler<*>) {
        val value = factory()
        val prevValue = handlersMap.putIfAbsent(context, value)
        if (prevValue == null) {
            Sync.lock(myOrderingsLock) {
                myHandlerOrder.add(value)
            }
        }
    }

    private fun <T : Any> ensureContextHandlerExists(context: RdContext<T>) {
        if (!handlersMap.containsKey(context)) {
            serializationCtx.serializers.register(RdContext.marshallerFor(context))
            doAddHandler(context) {
                if(context.heavy)
                    HeavySingleContextHandler(context, this)
                else
                    LightSingleContextHandler(context)
            }
        }
    }

    private fun preBindHandler(lifetime: Lifetime, handler: ISingleContextHandler<*>, key: String, proto: IProtocol) {
        if (handler !is HeavySingleContextHandler<*>) return

        handler.rdid = proto.identity.mix(rdid, key)
        handler.preBind(lifetime, this, key)
    }

    private fun bindHandler(handler: ISingleContextHandler<*>) {
        if (handler !is HeavySingleContextHandler<*>) return

        sendWithoutContexts {
            handler.bind()
        }
    }

    @Suppress("UNCHECKED_CAST")
    internal fun <T : Any> getContextHandler(context: RdContext<T>) : ISingleContextHandler<T> {
        return handlersMap[context] as ISingleContextHandler<T>
    }

    /**
     * Get a value set for a given context
     */
    @Suppress("UNCHECKED_CAST")
    fun <T : Any> getValueSet(context: RdContext<T>) : IAppendOnlyViewableConcurrentSet<T> {
        assert(context.heavy) { "Only heavy contexts have value sets, ${context.key} is not heavy" }
        return (getContextHandler(context) as HeavySingleContextHandler<T>).valueSet
    }

    /**
     * Registers a context to be used with this protocol. Must be invoked on protocol's scheduler
     */
    fun <T : Any> registerContext(context: RdContext<T>) {
        ensureContextHandlerExists(context)
    }

    private fun sendContextToRemote(context: RdContext<*>) {
        val proto = protocol ?: return

        sendWithoutContexts {
            proto.wire.send(rdid) { buffer ->
                RdContext.write(serializationCtx, buffer, context)
            }
        }
    }

    /**
     * Writes the current context values to the given buffer
     */
    fun writeCurrentMessageContext(buffer: AbstractBuffer) {
        if (isSendWithoutContexts) return writeEmptyContexts(buffer)

        // all handlers in handlersToWrite have been sent to the remote side
        val writtenSize = handlersToWrite.size
        buffer.writeShort(writtenSize.toShort())
        for (i in 0 until writtenSize) {
            handlersToWrite[i].writeValue(serializationCtx, buffer)
        }
    }

    internal fun readContext(buffer: AbstractBuffer): MessageContext {
        val numContextValues = buffer.readShort().toInt()
        val values = arrayOfNulls<Any>(numContextValues)
        val handlers = arrayOfNulls<ISingleContextHandler<*>>(numContextValues)
        assert(counterpartHandlers.size >= numContextValues) { "We know of ${counterpartHandlers.size} remote keys, received $numContextValues instead" }

        for (i in 0 until numContextValues) {
            val handler = counterpartHandlers[i]
            values[i] = handler.readValue(serializationCtx, buffer)
            handlers[i] = handler
        }

        return MessageContext(values as Array<Any>, handlers as Array<ISingleContextHandler<*>>)
    }

    internal class MessageContext(private val values: Array<Any>, private val handlers: Array<ISingleContextHandler<*>>) {
        inline fun update(action: () -> Unit) {
            val disposables = arrayOfNulls<AutoCloseable>(values.size)

            for (i in values.indices) {
                val handler = handlers[i].context as RdContext<Any>
                disposables[i] = handler.updateValue(values[i])
            }

            try {
                action()
            } finally {
                for (value in disposables) {
                    value?.close()
                }
            }
        }
    }

    companion object {
        /**
         * Writes an empty context
         */
        fun writeEmptyContexts(buffer: AbstractBuffer) {
            buffer.writeShort(0)
        }
    }

    override var async: Boolean
        get() = true
        set(value) = error("ProtocolContextHandler is always async")
}